kit/ChildSession.cpp

```diff
@@ -2484,23 +2486,51 @@ bool ChildSession::renderNextSlideLayer(SlideCompressor &scomp,
             uint64_t pixmapHash = hashSubBuffer(pixmap->data(), 0, 0, width, height, width, height) + getViewId();
             std::string json = jsonMsg;
             if (size_t start = json.find("%IMAGETYPE%"); start != std::string::npos)
-                json.replace(start, 11, "png");
+                json.replace(start, 11, "zstd");
             if (size_t start = json.find("%IMAGECHECKSUM%"); start != std::string::npos)
                 json.replace(start, 15, std::to_string(pixmapHash));
 
+            {
+                Poco::JSON::Parser parser;
+                Poco::JSON::Object::Ptr root = parser.parse(json).extract<Poco::JSON::Object::Ptr>();
+                root->set("width", width);
+                root->set("height", height);
+                std::stringstream ss;
+                root->stringify(ss);
+                json = ss.str();
+            }
             std::string response = "slidelayer: " + json;
 
             response += "\n";
 
-            output.reserve(response.size() + pixmap->size());
-            output.resize(response.size());
+            size_t compressed_max_size = ZSTD_COMPRESSBOUND(pixmap->size());
+            size_t max_required_size = response.size() + compressed_max_size;
+            output.resize(max_required_size);
             std::memcpy(output.data(), response.data(), response.size());
+            std::vector<char> compressedOutPut;
+            compressedOutPut.resize(ZSTD_COMPRESSBOUND(pixmap->size()));
+
+            if (tileMode == LibreOfficeKitTileMode::LOK_TILEMODE_BGRA)
+            {
+                png_row_info rowInfo;
+                rowInfo.rowbytes = pixmap->size();
+                // Following function just needs row size to transform from BGRA to RGBA
+                // We have a flat array so its safe to pass pixmap size as row size
+                Png::unpremultiply_bgra_data(nullptr, &rowInfo, pixmap->data());
```

Comment:
Ah - this is a pain; so I -think- we also do this in the web worker - or don't bother doing it at all because we create images with the pre-multiplied alpha hint in Chris' work. This unpremultiply is surprisingly expensive I'm afraid CPU-wise; it would be nice to rid ourselves of that too even if it's threaded. But we can leave that to a next-patch and/or easy-hack - can you file a github ticket with code pointers there ?